lib/phoenix_live_view/async_result.ex

```diff
@@ -0,0 +1,501 @@
+  """
+  def exit(%AsyncResult{} = result, reason) do
+    %AsyncResult{result | state: {:exit, reason}}
+  end
+
+  @doc """
+  Updates the state of the result to `:ok` and sets the result.
+
+  The `:ok?` field will also be set to `true` to indicate this result has
+  completed successfully at least once, regardless of future state changes.
+  """
+  def ok(%AsyncResult{} = result, value) do
+    %AsyncResult{result | state: :ok, ok?: true, result: value}
+  end
+
+  @doc """
+  Renders an async assign with slots for the different loading states.
+
+  ## Examples
+
+  ```heex
+  <AsyncResult.with_state :let={org} assign={@org}>
+    <:loading>Loading organization...</:loading>
+    <:empty>You don't have an organization yet</:error>
+    <:error :let={{_kind, _reason}}>there was an error loading the organization</:error>
+    <:canceled :let={_reason}>loading canceled</:canceled>
+    <%= org.name %>
+  <AsyncResult.with_state>
+  ```
+  """
+  attr :assign, :any, required: true
+  slot :loading
+
+  # TODO decide if we want an canceled slot
+  slot :canceled
+
+  # TODO decide if we want an empty slot
+  slot :empty,
+    doc:
+      "rendered when the result is loaded and is either nil or an empty list. Receives the result as a :let."
+
+  slot :failed,
+    doc:
+      "rendered when an error or exit is caught or assign_async returns `{:error, reason}`. Receives the error as a :let."
+
+  def with_state(assigns) do
+    case assigns.assign do
+      %AsyncResult{state: state, ok?: once_ok?, result: result} when state == :ok or once_ok? ->
+        if assigns.empty != [] && result in [nil, []] do
+          ~H|<%= render_slot(@empty, @assign.result) %>|
+        else
+          ~H|<%= render_slot(@inner_block, @assign.result) %>|
+        end
+
+      %AsyncResult{state: :loading} ->
+        ~H|<%= render_slot(@loading) %>|
+
+      %AsyncResult{state: {:error, {:canceled, reason}}} ->
+        if assigns.canceled != [] do
+          assigns = Phoenix.Component.assign(assigns, reason: reason)
+          ~H|<%= render_slot(@canceled, @reason) %>|
+        else
+          ~H|<%= render_slot(@failed, @assign.state) %>|
+        end
+
+      %AsyncResult{state: {kind, _reason}} when kind in [:error, :exit] ->
+        ~H|<%= render_slot(@failed, @assign.state) %>|
+    end
+  end
+
+  @doc """
+  Assigns keys asynchronously.
+
+  The task is linked to the caller and errors are wrapped.
+  Each key passed to `assign_async/3` will be assigned to
+  an `%AsyncResult{}` struct holding the status of the operation
+  and the result when completed.
+  """
+  def assign_async(%Socket{} = socket, key_or_keys, func)
+      when (is_atom(key_or_keys) or is_list(key_or_keys)) and
+             is_function(func, 0) do
+    keys = List.wrap(key_or_keys)
+
+    # verifies result inside task
+    wrapped_func = fn ->
+      case func.() do
+        {:ok, %{} = assigns} ->
+          if Map.keys(assigns) -- keys == [] do
```

Comment:
```suggestion
          if missing_key = Enum.find(keys, &not is_map_key(assigns, & &1)) do
```

Reply:
Currently we don't check if we return more keys than the ones specified, is that ok?